src/v/kafka/server/handlers/fetch.cc

```diff
@@ -672,6 +673,482 @@ class parallel_fetch_plan_executor final : public fetch_plan_executor::impl {
+    query_requests(std::vector<ntp_fetch_config> requests) {
+        // The last visible indexes need to be populated before partitions
+        // are read. If they are populated afterwards then the
+        // last_visible_index could be updated after the partition is read,
+        // but before the index is queried. This leads to a race condition
+        // where we're waiting for the last_visible_index to change even
+        // though the partition has data we haven't read.
+        //
+        // Note that `last_visible_index` is used instead of the
+        // `commit_offset` as the commit offset won't change when partitions
+        // are produced to without acks=all. The `last_visible_index` more
+        // closely corresponds to the Kafka high watermark as well.
+        std::vector<model::offset> last_visible_indexes{requests.size()};
+        std::vector<std::tuple<size_t, model::partition_id>> errored_partitions;
+        size_t total_size{0};
+        bool has_error{false};
+
+        for (size_t i = 0; i < requests.size(); i++) {
+            const auto& req = requests[i];
+            auto part = _ctx.mgr.get(req.ktp());
+            if (!part) {
+                errored_partitions.emplace_back(i, req.ktp().get_partition());
+                continue;
+            }
+            auto consensus = part->raft();
+            if (!consensus) {
+                errored_partitions.emplace_back(i, req.ktp().get_partition());
+                continue;
+            }
+            last_visible_indexes[i] = consensus->last_visible_index();
+        }
+
+        // A read_result needs to be returned for every partition. Hence,
+        // the function can't return before calling
+        // `fetch_ntps_in_parallel`.
+        std::vector<read_result> results = co_await fetch_ntps_in_parallel(
+          _ctx.mgr,
+          _ctx.srv.get_replica_selector(),
+          std::move(requests),
+          _ctx.foreign_read,
+          _ctx.deadline,
+          _ctx.bytes_left,
+          _ctx.srv.memory(),
+          _ctx.srv.memory_fetch_sem());
+
+        // If we weren't able to read the last_visible_index for a partition
+        // before calling `fetch_ntps_in_parallel` then we need to
+        // return with an error for that partition.
+        for (auto [i, partition] : errored_partitions) {
+            results[i] = read_result(error_code::not_leader_for_partition);
+            results[i].partition = partition;
+        }
+
+        for (const auto& r : results) {
+            total_size += r.data_size_bytes();
+            if (r.error != error_code::none) {
+                has_error = true;
+            }
+        }
+
+        co_return query_results{
+          .last_visible_indexes = std::move(last_visible_indexes),
+          .has_error = has_error,
+          .results = std::move(results),
+          .total_size = total_size,
+        };
+    }
+
+    std::optional<size_t> register_waiters(const auto& request_indexes) {
```

Comment:
This returns an optional<size_t>, what does it mean?

Reply:
optional size_t (brackets eaten by html)

Reply:
Overall please document the semantics of this method. It seems like it returns the first failure index, but also may also register some offset waiters depending on where it detects the error.

Reply:
Also since there are two types of indexes here, the original domain (suitable to index _ctx.requests) and the remapped domain (suitable to index `reuqest_indexes`) we should be clear which it is (the former, here).

Reply:
Added documentation to this method and clarified which index is involved in it.

Reply:
Looks, good, thanks!